public/client/TracyCallstack.cpp

```diff
@@ -90,9 +90,111 @@ extern "C" const char* ___tracy_demangle( const char* mangled )
 #endif
 #endif
 
+#if TRACY_HAS_CALLSTACK == 3
+#   define TRACY_USE_IMAGE_CACHE
+#   include <link.h>
+#endif
+
 namespace tracy
 {
 
+#ifdef TRACY_USE_IMAGE_CACHE
+// when we have access to dl_iterate_phdr(), we can build a cache of address ranges to image paths
+// so we can quickly determine which image an address falls into.
+// We refresh this cache only when we hit an address that doesn't fall into any known range.
+class ImageCache
+{
+public:
+    struct ImageEntry
+    {
+        ImageEntry( void* startAddress, void* endAddress, const char* name )
+        : m_startAddress( startAddress ), m_endAddress( endAddress ), m_name( name ) {}
+
+        void* m_startAddress;
+        void* m_endAddress;
+        const char* m_name;
+    };
+
+    ImageCache()
+    {
+        m_images = (tracy::FastVector<ImageEntry>*)tracy::tracy_malloc( sizeof( tracy::FastVector<ImageEntry> ) );
+        new(m_images) tracy::FastVector<ImageEntry>( 512 );
+
+        Refresh();
+    }
+    const ImageEntry* GetImageForAddress( void* address ) 
+    {
+        const ImageEntry* entry = GetImageEntryForAddress( address );
+        if( !entry )
+        {
+            Refresh();
+            return GetImageEntryForAddress( address );
+        }
+        return entry;
+    }
+
+
+private:
+    tracy::FastVector<ImageEntry>* m_images;
+    const char* m_imageName = nullptr;
+
+    static int Callback( struct dl_phdr_info* info, size_t size, void* data ) 
+    {
+        ImageCache* cache = reinterpret_cast<ImageCache*>( data );
+
+        ImageEntry* image = cache->m_images->push_next();
+        image->m_startAddress = reinterpret_cast<void*>( info->dlpi_addr );
+        const uint32_t headerCount = info->dlpi_phnum;
+        assert( headerCount > 0);
+        image->m_endAddress = reinterpret_cast<void*>( info->dlpi_addr + 
+            info->dlpi_phdr[info->dlpi_phnum - 1].p_vaddr + info->dlpi_phdr[info->dlpi_phnum - 1].p_memsz);
+
+        // the base executable name isn't provided when iterating with dl_iterate_phdr,
+        // so we must get it in an alternative way and cache it
+        if( info->dlpi_name && info->dlpi_name[0] != '\0' )
+        {
+            image->m_name = info->dlpi_name;
+        }
+        else
+        {
+            if( !cache->m_imageName )
+            {
+                Dl_info dlInfo;
+                if( dladdr( (void *)info->dlpi_addr, &dlInfo ) )
+                {
+                    cache->m_imageName = dlInfo.dli_fname;
+                }
+            }
+            image->m_name = cache->m_imageName;
+        }
```

Comment:
Is `image->m_name` valid after this function returns? My guess would be that the `info` argument is only valid inside this function, and the `dlInfo` provided filename will be invalidated after the next call to `dladdr`.

Reply:
good point - fixed.

Reply:
Where is this memory released?